crates/uv-distribution/src/pyproject_mut.rs

```diff
@@ -194,19 +197,45 @@ fn implicit() -> Item {
 }
 
 /// Adds a dependency to the given `deps` array.
-pub fn add_dependency(req: &Requirement, deps: &mut Array) {
+pub fn add_dependency(req: Requirement, deps: &mut Array, has_source: bool) -> Result<(), Error> {
     // Find matching dependencies.
-    let to_replace = find_dependencies(&req.name, deps);
-    if to_replace.is_empty() {
-        deps.push(req.to_string());
-    } else {
-        // Replace the first occurrence of the dependency and remove the rest.
-        deps.replace(to_replace[0], req.to_string());
-        for &i in to_replace[1..].iter().rev() {
-            deps.remove(i);
+    let mut to_replace = find_dependencies(&req.name, deps);
+    match to_replace.as_slice() {
+        [] => deps.push(req.to_string()),
+        [_] => {
+            let (i, mut old_req) = to_replace.remove(0);
+            update_requirement(&mut old_req, req, has_source);
+            deps.replace(i, old_req.to_string());
         }
+        // Cannot perform ambiguous updates.
+        _ => return Err(Error::Ambiguous),
     }
     reformat_array_multiline(deps);
+    Ok(())
+}
+
+/// Update an existing requirement.
+fn update_requirement(old: &mut Requirement, new: Requirement, has_source: bool) {
+    // Add any new extras.
+    old.extras.extend(new.extras);
+    old.extras.dedup();
```

Comment:
Do we need to sort?

Reply:
I don't think so? This is updating `pyproject.toml` not the lockfile.

Reply:
But `dedup` only deduplicates adjacent entries, right?

Reply:
Oh right, fixed. I wonder if `extras` should be a `HashSet`..